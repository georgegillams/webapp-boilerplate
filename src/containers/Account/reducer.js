import produce from 'immer';

import { loadShowcases } from './actions';

export const initialState = {
  loadingShowcases: false,
  loadShowcasesError: null,
  showcases: [],
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case loadShowcases.REQUEST:
        draft.loadingShowcases = true;
        draft.loadShowcasesError = null;
        break;

      case loadShowcases.SUCCESS:
        draft.loadingShowcases = false;
        draft.showcases = payload.data;
        break;

      case loadShowcases.FAILURE:
        draft.loadingShowcases = false;
        draft.loadShowcasesError = payload;
        break;
    }
  });

export default reducer;
