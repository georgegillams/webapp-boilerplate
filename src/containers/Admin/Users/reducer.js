import produce from 'immer';

import { load, remove, requestMagicLink } from './actions';

export const initialState = {
  users: null,
  loading: false,
  loadError: null,

  userToRemove: null,
  removing: false,
  removeError: null,

  userToLogIn: null,
  requesting: false,
  requestError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.users = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;

      case remove.TRIGGER:
        draft.userToRemove = payload;
        break;

      case remove.REQUEST:
        draft.removing = true;
        draft.removeError = null;
        break;

      case remove.SUCCESS:
        draft.removing = false;
        break;

      case remove.FAILURE:
        draft.removing = false;
        draft.removeError = payload;
        break;

      case requestMagicLink.TRIGGER:
        draft.userToLogIn = payload;
        break;

      case requestMagicLink.REQUEST:
        draft.requesting = true;
        draft.requestError = null;
        break;

      case requestMagicLink.SUCCESS:
        draft.requesting = false;
        break;

      case requestMagicLink.FAILURE:
        draft.requesting = false;
        draft.requestError = payload;
        break;
    }
  });

export default reducer;
