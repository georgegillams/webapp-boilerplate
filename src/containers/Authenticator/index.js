import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { setCookiesAllowed, loadAuth } from './actions';
import { selectState } from './selectors';
import Authenticator from './Container';

const mapStateToProps = createStructuredSelector({
  authenticatorState: selectState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    setCookiesAllowed: payload => dispatch(setCookiesAllowed(payload)),
    loadAuth: payload => dispatch(loadAuth(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo)(Authenticator);
