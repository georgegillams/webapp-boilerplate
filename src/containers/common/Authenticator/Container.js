import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import DebugObject from 'components/common/DebugObject';

const Authenticator = props => {
  const { loadAuth, authenticatorState } = props;

  useEffect(() => {
    loadAuth();
  }, []);

  return (
    <div>
      <DebugObject
        debugTitle="Authenticator"
        debugObject={{
          loadAuth,
          authenticatorState,
        }}
      />
    </div>
  );
};

Authenticator.propTypes = {
  loadAuth: PropTypes.func.isRequired,
  authenticatorState: PropTypes.shape({
    loadingAuth: PropTypes.bool,
    loadAuthError: PropTypes.object,
    user: PropTypes.object,
  }).isRequired,
};

Authenticator.defaultProps = {};

export default Authenticator;
