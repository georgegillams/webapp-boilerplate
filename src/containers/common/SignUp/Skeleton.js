import React from 'react';
import PageContainer from 'components/common/PageContainer';

import Skeleton, { SKELETON_STYLES } from '@george-gillams/components/skeleton';

const PageSkeleton = props => {
  return (
    <PageContainer {...props}>
      <Skeleton skeletonStyle={SKELETON_STYLES.section} />
      <Skeleton skeletonStyle={SKELETON_STYLES.input} style={{ marginTop: '1.2rem' }} />
      <Skeleton skeletonStyle={SKELETON_STYLES.button} />
    </PageContainer>
  );
};

PageSkeleton.propTypes = {};

PageSkeleton.defaultProps = {};

export default PageSkeleton;
