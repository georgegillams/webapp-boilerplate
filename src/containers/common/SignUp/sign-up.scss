@import '~@george-gillams/webapp/utils/tokens';

.sign-up {
  &__link {
    margin-top: $spacing-base;
  }

  &__form {
    margin-top: $spacing-base;
  }
}
