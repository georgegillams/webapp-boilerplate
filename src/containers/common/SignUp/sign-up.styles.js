import styled from 'styled-components';
import { SignUpForm } from 'components/common/Forms';
import { spacingBase, spacingLg } from '@george-gillams/components/constants/layout';
import TextLink from 'components/common/TextLink';

export const StyledSignUpForm = styled(SignUpForm)`
  margin-top: ${spacingLg};
`;

export const StyledTextLink = styled(TextLink)`
  margin-top: ${spacingBase};
`;
