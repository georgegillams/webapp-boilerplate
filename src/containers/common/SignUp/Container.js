import React, { useState } from 'react';
import PropTypes from 'prop-types';
import TextLink from 'components/common/TextLink';
import PageTitle from 'components/common/PageTitle';
import DebugObject from 'components/common/DebugObject';
import LoadingCover from '@george-gillams/components/loading-cover';

import Skeleton from './Skeleton';

import { LoggedOutOnly } from 'components/common/Walls';
import { SignUpForm } from 'components/common/Forms';
import CookiesRequired from 'containers/common/CookiesRequired';
import { CONSENT_STATE_ALLOWED } from 'containers/common/Consent/constants';
import STYLES from './sign-up.scss';
import { cssModules } from '@george-gillams/components/helpers/cssModules';
import PageContainer from 'components/common/PageContainer';

const getClassName = cssModules(STYLES);

const SignUp = props => {
  const [credentials, setCredentials] = useState({});

  const {
    signUp,

    consentState,
    signUpState,
    authenticatorState,
  } = props;

  let preSubmitText = null;
  if (signUpState.signUpResult && signUpState.signUpResult.success) {
    preSubmitText = signUpState.signUpResult.success;
  }
  if (signUpState.signUpError && signUpState.signUpError.errorMessage) {
    preSubmitText = signUpState.signUpError.errorMessage;
  }

  // We shouldn't allow a user to sign up unless cookies are consented to
  const page = (
    <PageContainer bottomPadding>
      <CookiesRequired reason={'sign up'} />
      <LoggedOutOnly user={authenticatorState.user}>
        <PageTitle name="Sign up">
          <SignUpForm
            className={getClassName('sign-up__form')}
            disabled={consentState.cookieConsent !== CONSENT_STATE_ALLOWED}
            loading={signUpState.signingUp}
            credentials={credentials}
            onDataChanged={setCredentials}
            onSubmit={() => {
              if (consentState.cookieConsent === CONSENT_STATE_ALLOWED) {
                signUp(credentials);
              }
            }}
            preSubmitText={preSubmitText}
          />
          <TextLink className={getClassName('sign-up__link')} href="/login">
            Already got an account? Login here.
          </TextLink>
        </PageTitle>
      </LoggedOutOnly>
    </PageContainer>
  );

  return (
    <>
      <LoadingCover loadingSkeleton={Skeleton} loading={authenticatorState.user === undefined} error={false}>
        {page}
      </LoadingCover>
      <DebugObject
        debugTitle="SignUp"
        debugObject={{
          signUp,
          authenticatorState,
          signUpState,
        }}
      />
    </>
  );
};

SignUp.propTypes = {
  signUp: PropTypes.func.isRequired,
  signUpState: PropTypes.shape({
    signingUp: PropTypes.bool,
    signUpError: PropTypes.object,
    signUpResult: PropTypes.object,
  }).isRequired,
  authenticatorState: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  consentState: PropTypes.shape({
    cookieConsent: PropTypes.string,
  }).isRequired,
};

SignUp.defaultProps = {
  authenticatorState: null,
  signUpState: null,
};

export default SignUp;
