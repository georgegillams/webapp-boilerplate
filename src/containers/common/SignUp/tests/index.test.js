import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from '../reducer';
import { initialState as initialAuthenticatorState } from 'containers/common/Authenticator/reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import SignUpIndex from '../index';
import SignUp from '../Container';

describe('<SignUp />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUpIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signUpError', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signUpError: { error: 'not_found', errorMessage: 'No matching email found' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signUpResult', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signUpResult: { uname: 'uname', email: 'email' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signingUp=true', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signingUp: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with user', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
