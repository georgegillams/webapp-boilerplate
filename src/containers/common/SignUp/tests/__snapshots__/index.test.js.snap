// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> should render correctly - index 1`] = `
<div>
  <div
    aria-label="Loading"
    class="sc-dIfARi evaMPU page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-bjfHbI ibzWhy"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI izPJkq"
      style="margin-top: 1.2rem;"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI beSZFO"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with initial state 1`] = `
<div>
  <div
    aria-label="Loading"
    class="sc-dIfARi evaMPU page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-bjfHbI ibzWhy"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI izPJkq"
      style="margin-top: 1.2rem;"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI beSZFO"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signUpError 1`] = `
<div>
  <div
    aria-label="Loading"
    class="sc-dIfARi evaMPU page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-bjfHbI ibzWhy"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI izPJkq"
      style="margin-top: 1.2rem;"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI beSZFO"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signUpResult 1`] = `
<div>
  <div
    aria-label="Loading"
    class="sc-dIfARi evaMPU page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-bjfHbI ibzWhy"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI izPJkq"
      style="margin-top: 1.2rem;"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI beSZFO"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signingUp=true 1`] = `
<div>
  <div
    aria-label="Loading"
    class="sc-dIfARi evaMPU page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-bjfHbI ibzWhy"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI izPJkq"
      style="margin-top: 1.2rem;"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="sc-bjfHbI beSZFO"
    >
      <div
        class="sc-idXgbr jlSlT"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with user 1`] = `
<div>
  <div
    class="sc-dIfARi bvVxIG page-containerstyles__StyledPageContainer-sc-19l7k4-0"
    width="default"
  >
    <div
      class="sc-ksBlkl gDlyuV"
    >
      <h1
        class="sc-gswNZR gzqAHo sc-fEXmlR cIHzsJ"
        id="you're-logged-in"
      >
        You're logged in
      </h1>
      <span
        class="sc-hHTYSt cRxHgE"
      >
        <a
          class="sc-iBYQkv jNlLug"
          href="/account"
        >
          Go to your account.
        </a>
      </span>
    </div>
  </div>
</div>
`;
